homeassistant/components/tractive/sensor.py

```diff
@@ -0,0 +1,184 @@
+from homeassistant.const import (
+    ATTR_BATTERY_LEVEL,
+    DEVICE_CLASS_BATTERY,
+    PERCENTAGE,
+    TIME_MINUTES,
+)
+from homeassistant.core import callback
+from homeassistant.helpers.dispatcher import async_dispatcher_connect
+
+from .const import (
+    ATTR_DAILY_GOAL,
+    ATTR_MINUTES_ACTIVE,
+    DOMAIN,
+    SERVER_UNAVAILABLE,
+    TRACKER_ACTIVITY_STATUS_UPDATED,
+    TRACKER_HARDWARE_STATUS_UPDATED,
+)
+from .entity import TractiveEntity
+
+
+@dataclass
+class TractiveSensorEntityDescription(SensorEntityDescription):
+    """Class describing Tractive sensor entities."""
+
+    event_type: str | None = None
+    attributes: tuple = ()
+
+
+ATTR_ACTIVITY = "activity"
+ATTR_HARDWARE = "hardware"
+
+SENSOR_TYPES = (
+    TractiveSensorEntityDescription(
+        key=ATTR_BATTERY_LEVEL,
+        name="Battery Level",
+        native_unit_of_measurement=PERCENTAGE,
+        device_class=DEVICE_CLASS_BATTERY,
+        event_type=ATTR_HARDWARE,
+    ),
+    TractiveSensorEntityDescription(
+        key=ATTR_MINUTES_ACTIVE,
+        name="Minutes Active",
+        icon="mdi:clock-time-eight-outline",
+        native_unit_of_measurement=TIME_MINUTES,
+        event_type=ATTR_ACTIVITY,
+        attributes=(ATTR_DAILY_GOAL,),
+    ),
+)
+
+
+async def async_setup_entry(hass, entry, async_add_entities):
+    """Set up Tractive device trackers."""
+    client = hass.data[DOMAIN][entry.entry_id]
+
+    trackables = await client.trackable_objects()
+
+    entities = []
+
+    for item in trackables:
+        trackable = await item.details()
+        tracker = client.tracker(trackable["device_id"])
+        tracker_details = await tracker.details()
+        for description in SENSOR_TYPES:
+            unique_id = f"{trackable['_id']}_{description.key}"
+            if description.event_type == ATTR_HARDWARE:
+                entities.append(
+                    TractiveHardwareSensor(
+                        client.user_id,
+                        trackable,
+                        tracker_details,
+                        unique_id,
+                        description,
+                    )
+                )
+            if description.event_type == ATTR_ACTIVITY:
+                entities.append(
+                    TractiveActivitySensor(
+                        client.user_id,
+                        trackable,
+                        tracker_details,
+                        unique_id,
+                        description,
+                    )
+                )
+
+    async_add_entities(entities)
+
+
+class TractiveSensor(TractiveEntity, SensorEntity):
+    """Tractive sensor."""
+
+    def __init__(self, user_id, trackable, tracker_details, unique_id, description):
+        """Initialize sensor entity."""
+        super().__init__(user_id, trackable, tracker_details)
+
+        self._attr_unique_id = unique_id
+        self._attr_device_info = self._device_info
+        self.entity_description = description
```

Comment:
Make it private?
```
self._entity_description = description
```

Reply:
This can't be private, it needs to override the base class

Reply:
Sorry, I don't really get it, the base class `TractiveEntity` does not have `entity_description`, and even if it had, nothing prevents it from making it private, there is no access to this attribute outside of the `self` context

Reply:
`SensorEntity` has it, and `SensorEntity` uses `self.entity_description` to set properties correctly (same with `Entity` which `SensorEntity` inherits)

Reply:
Ahhh, I didn't know it's a part of the generic interface, sorry.